### PR TITLE
MBS-11962: Update privileges for all database users on schema changes

### PR DESCRIPTION
**Pull request by mwiencek**
# Problem

MBS-11962

When new tables or schemas are added, production database users like `musicbrainz_ro` may not be granted access to them.

# Solution

Adds a new script, admin/UpdateDatabasePrivileges.pl, which is invoked from upgrade.sh and updates privileges for the following users if they exist:

 * `musicbrainz_ro`
 * `caa_redirect`
 * `sir`

The script works by revoking all write privileges from `musicbrainz_ro`, and then granting `USAGE` on all schemas, and `SELECT` privileges on all tables in those schemas.

Next, for the last two schemas, it revokes all privileges from them (including `SELECT` and `USAGE`), and then simply grants them membership in the `musicbrainz_ro` role.

# Testing

I added the three roles in question to my local database and ran admin/UpdateDatabasePrivileges.pl. I observed that each of the three users could `SELECT` from any table in our schemas, but not modify them.

### CODE REVIEW

File: admin/UpdateDatabasePrivileges.pl [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+        );
+    }
+    for my $role (qw( caa_redirect sir )) {
+        next unless $role_names{$role};
+
+        my $quoted_role = $sql->dbh->quote_identifier($role);
+        for my $schema (@FULL_SCHEMA_LIST) {
+            my $quoted_schema = $sql->dbh->quote_identifier($schema);
+            $sql->do(
+                "REVOKE SELECT, $non_select_privileges " .
```
Comment (reosarevok) [inline admin/UpdateDatabasePrivileges.pl:49]: Would it work to just do `REVOKE ALL PRIVILEGES`?

File: admin/UpdateDatabasePrivileges.pl [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+use MusicBrainz::Server::Constants qw( @FULL_SCHEMA_LIST );
+use MusicBrainz::Server::Context;
+
+my $c = MusicBrainz::Server::Context->create_script_context(
+    database => 'SYSTEM_MAINTENANCE',
+);
+my $sql = $c->sql;
+
+my $non_select_privileges =
+    'INSERT, UPDATE, DELETE, TRUNCATE, REFERENCES, TRIGGER';
```
Comment (reosarevok) [inline admin/UpdateDatabasePrivileges.pl:18]: Something in my brain wants this to be a `qw()` and a `join` but this is probably fine. That said, if you're going to manually `GRANT SELECT` after the `REVOKE`, why not just revoke everything anyway?

File: admin/InitDb.pl [hunk 3]
```diff
@@ -238,18 +239,7 @@ sub CreateRelations
...
-            documentation
-            event_art_archive
-            json_dump
-            report
-            sitemaps
-            statistics
-            wikidocs
-            dbmirror2
-        ));
+        for @FULL_SCHEMA_LIST;
```
Comment (reosarevok) [inline admin/InitDb.pl:242]: I'm going to channel @yvanzo and say that this should probably be a separate commit than the new script?

**Review by reosarevok — COMMENTED** [review]
The idea seems great. Can we add a test for the script, please? :)

File: admin/UpdateDatabasePrivileges.pl (hunk 2 above)
Comment (reosarevok) [reply admin/UpdateDatabasePrivileges.pl]: Ah, ok, didn't realize there's a lot more privileges than these.

**Comment by reosarevok** [comment]
Tests seem to be failing

**Comment by mwiencek** [comment]
> Tests seem to be failing

Seems to be because the `WITH INHERIT` clause I'm using was added in PostgreSQL 16.

**Review by reosarevok — APPROVED** [review]
LGTMBDNT

File: admin/UpdateDatabasePrivileges.pl [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    }
+    for my $other_role (@other_ro_roles) {
+        unless ($existing_role_names{$other_role}) {
+            log_info {
+                qq(The role "$other_role" does not exist; skipping.)
+            };
+            next;
+        }
+
+        my $quoted_other_role = $sql->dbh->quote_identifier($other_role);
```
Comment (yvanzo) [inline admin/UpdateDatabasePrivileges.pl:107]: It would be helpful to log roles whose database privileges are being updated, and in which database.

File: t/script/UpdateDatabasePrivileges.t [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+    );
+
+    for my $role (@roles) {
+        my $conn = Databases->get_connection("db_priv_test_${role}");
+        my $sql = Sql->new($conn->conn);
+
+        $sql->auto_commit;
+        like exception {
+            $sql->do('SELECT * FROM musicbrainz.db_priv_test;');
+        }, qr/permission denied for schema musicbrainz/;
```
Comment (yvanzo) [inline t/script/UpdateDatabasePrivileges.t:50]: Those tests should be documented through their output.

File: admin/UpdateDatabasePrivileges.pl [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+     WHERE rolname = any(?)
+    SQL
+    [$primary_ro_role, @other_ro_roles],
+);
+my %existing_role_names = map { $_ => 1 } @$existing_role_names;
+
+if ($existing_role_names{$primary_ro_role}) {
+    my $quoted_primary_role = $sql->dbh->quote_identifier($primary_ro_role);
+    $sql->begin;
+    for my $schema (@FULL_SCHEMA_LIST) {
```
Comment (yvanzo) [inline admin/UpdateDatabasePrivileges.pl:69]: There should probably be a condition for each schema has it might not exist in the current database. (?)

**Review by yvanzo — COMMENTED** [review]
I created the roles `musicbrainz_ro`, `caa_redirect` and `sir` using `admin/psql` then ran the new script and obtained the following error message:

```
Failed query:
        'GRANT "musicbrainz_ro"
   TO "caa_redirect"
 WITH INHERIT TRUE;
'
        ()
42601 DBD::Pg::st execute failed: ERROR:  syntax error at or near "INHERIT"
LINE 3:  WITH INHERIT TRUE;
              ^ [for Statement "GRANT "musicbrainz_ro"
   TO "caa_redirect"
 WITH INHERIT TRUE;
"]
 at /musicbrainz-server/admin/../lib/Sql.pm line 123.
        Sql::catch {...} (MusicBrainz::Server::Exceptions::DatabaseError=HASH(0x5acb49453658)) called at /musicbrainz-server/perl_modules/lib/perl5/Try/Tiny.pm line 123
        Try::Tiny::try(CODE(0x5acb49453dd8), Try::Tiny::Catch=REF(0x5acb49353978)) called at /musicbrainz-server/admin/../lib/Sql.pm line 124
        Sql::do(Sql=HASH(0x5acb48ebf288), "GRANT \"musicbrainz_ro\"\x{a}   TO \"caa_redirect\"\x{a} WITH INHERIT TRUE;\x{a}") called at ./admin/UpdateDatabasePrivileges.pl line 119

```

**Comment by mwiencek** [comment]
Thanks!

> Last, when running this script (without --nogrant) two times in a row, two additional lines are logged:

Yeah, those are logged by PG if you try to grant a privilege that already exists. It's harmless though and may be useful information (if it wasn't expected for them to exist).
